verifier/src/channel.rs

```diff
@@ -191,9 +194,26 @@ where
         let queries = self.trace_queries.take().expect("already read");
 
         // make sure the states included in the proof correspond to the trace commitment
+        let items: Vec<H::Digest> = if self.num_partitions == 1 {
+            queries.main_states.rows().map(|row| H::hash_elements(row)).collect()
+        } else {
+            let number_of_base_field_elements = queries.main_states.num_columns();
+            let num_elements_per_partition =
+                number_of_base_field_elements.div_ceil(self.num_partitions);
+            let mut buffer = vec![H::Digest::default(); self.num_partitions];
+
+            queries
+                .main_states
+                .rows()
+                .map(|row| {
+                    row.chunks(num_elements_per_partition)
+                        .zip(buffer.iter_mut())
+                        .for_each(|(chunk, buf)| *buf = H::hash_elements(chunk));
+                    H::merge_many(&buffer)
+                })
+                .collect()
+        };
```

Comment:
I would probably create a helper function to handle this and code on line 206 - 243. It could look something like:
```rust
fn hash_row<H, E>(row: &[E], num_partitions: usize) -> H::Digest
```
It would be slightly less efficient than this implementation - but I think the impact on the verifier would be negligible.

Reply:
Added